### PR TITLE
Task/WP-157: View Submitter Users (New Admin Page)

### PR DESCRIPTION
**Pull request by jmcmillenmusic**
## Overview

APCD admins should be able to view the records of all submitter users including basic information (names, emails, etc.) and payor codes and be able to update names and emails if needed.

## Related

- [WP-157](https://tacc-main.atlassian.net/browse/WP-157)

## Changes

The Admin dropdown in the navigation bar now contains the View Submitter Users page, which displays a table containing submitter ID, user ID, user number, user email, user name, payor code, and an Actions column of dropdown menu. These dropdown menus allow an admin to view and edit records for that specific submitter user. Only user name and user email can be edited.

## Testing

1. Click on Admin, then View Submitter Users. The table of submitter users should load.
2. Click on a user's Select Action dropdown and select View Record.
3. A modal should appear displaying the submitter user's information. Click the X to close it.
4. Click on a user's Select Action dropdown and select Edit Record.
5. Edit the user's name and/or email and click Submit. The user's information should be updated.

## UI

image

### CODE REVIEW

File: apcd_cms/src/apps/view_submitter_users/templates/view_submitter_user_edit_modal.html
Comment (chandra-tacc) [inline apcd_cms/src/apps/view_submitter_users/templates/view_submitter_user_edit_modal.html:1]: can be deleted - unused. Modal is 100% react

File: apcd_cms/src/apps/view_submitter_users/templates/view_submitter_user_modal.html
Comment (chandra-tacc) [inline apcd_cms/src/apps/view_submitter_users/templates/view_submitter_user_modal.html:1]: can be deleted - unused. Modal is 100% react

File: apcd_cms/src/apps/view_submitter_users/templates/view_user_edit_success.html
Comment (chandra-tacc) [inline apcd_cms/src/apps/view_submitter_users/templates/view_user_edit_success.html:1]: can be deleted - unused. Modal is 100% react

File: apcd_cms/src/apps/view_submitter_users/templates/view_user_edit_error.html
Comment (chandra-tacc) [inline apcd_cms/src/apps/view_submitter_users/templates/view_user_edit_error.html:1]: can be deleted - unused. Modal is 100% react

File: apcd_cms/src/client/src/hooks/admin/index.ts [hunk 1]
```diff
@@ -111,11 +114,32 @@ export type ExceptionResult = {
...
+  user_number: string;
+  created_at: string;
+  updated_at: string;
+  user_email: string;
+  user_name: string;
+  payor_code: string;
+};
+
+export type SubmitterUserResult = {
+  // selected_status: string;
```
Comment (chandra-tacc) [inline apcd_cms/src/client/src/hooks/admin/index.ts:129]: can be removed

File: apcd_cms/src/apps/utils/apcd_database.py [hunk 2]
```diff
@@ -152,6 +152,44 @@ def get_user_role(user):
...
+    conn = None
+    try:
+        conn = psycopg.connect(
+            host=APCD_DB['host'],
+            dbname=APCD_DB['database'],
+            user=APCD_DB['user'],
+            password=APCD_DB['password'],
+            port=APCD_DB['port'],
+            sslmode='require'
+        )
```
Comment (chandra-tacc) [inline apcd_cms/src/apps/utils/apcd_database.py:166]:
```suggestion
        conn = db_connect()
```

File: apcd_cms/src/apps/utils/apcd_database.py [hunk 3]
```diff
@@ -152,6 +152,44 @@ def get_user_role(user):
...
+        submitter_users.user_number,
+        users.user_email,
+        users.user_name,
+        submissions.payor_code
+        FROM submitter_users
+        JOIN users
+           ON submitter_users.user_id = users.user_id
+           AND submitter_users.user_number = users.user_number
+        JOIN submissions
+            ON submitter_users.submitter_id = submissions.submitter_id;
```
Comment (chandra-tacc) [inline apcd_cms/src/apps/utils/apcd_database.py:179]: having a order by helps the listing view consistent on refreshes
```suggestion
            ON submitter_users.submitter_id = submissions.submitter_id
            ORDER BY users.user_id;
```

File: apcd_cms/src/apps/utils/apcd_database.py (hunk 2 above)
Comment (chandra-tacc) [reply apcd_cms/src/apps/utils/apcd_database.py:166]: convenient util to setup connection

File: apcd_cms/src/apps/view_submitter_users/views.py
Comment (chandra-tacc) [inline apcd_cms/src/apps/view_submitter_users/views.py:1]: There is lot of code duplication with view_users, if the only thing is the underlying db function used, can we make this better ?

File: apcd_cms/src/client/src/components/Submitter/ViewSubmitterUsers/ViewSubmitterUsers.tsx
Comment (chandra-tacc) [inline apcd_cms/src/client/src/components/Submitter/ViewSubmitterUsers/ViewSubmitterUsers.tsx:1]: We could make ViewUsers a shared component with properties that distinguish between submitter vs all users. Things that vary:
1) hook used
2) view and edit modal content.

If the data is different (column names), could we make it so, we pick the columns dynamically based on submitter vs all users.

If this is too much for this PR, feel free to open a follow up task

**Review by chandra-tacc — CHANGES_REQUESTED** [review]

File: apcd_cms/src/apps/view_submitter_users/templates/view_submitter_user_edit_modal.html
Comment (jmcmillenmusic) [reply apcd_cms/src/apps/view_submitter_users/templates/view_submitter_user_edit_modal.html:1]: I just deleted this file.

File: apcd_cms/src/apps/view_submitter_users/templates/view_submitter_user_modal.html
Comment (jmcmillenmusic) [reply apcd_cms/src/apps/view_submitter_users/templates/view_submitter_user_modal.html:1]: I just deleted this file.

File: apcd_cms/src/apps/view_submitter_users/templates/view_user_edit_success.html
Comment (jmcmillenmusic) [reply apcd_cms/src/apps/view_submitter_users/templates/view_user_edit_success.html:1]: I just deleted this file.

File: apcd_cms/src/apps/view_submitter_users/templates/view_user_edit_error.html
Comment (jmcmillenmusic) [reply apcd_cms/src/apps/view_submitter_users/templates/view_user_edit_error.html:1]: I just deleted this file.

File: apcd_cms/src/client/src/hooks/admin/index.ts (hunk 1 above)
Comment (jmcmillenmusic) [reply apcd_cms/src/client/src/hooks/admin/index.ts:129]: Remove the commented-out portions or _everything_ under `export type SubmitterUserResult`?

Comment (chandra-tacc) [reply apcd_cms/src/client/src/hooks/admin/index.ts]: no, just the commented out code, just unnecessary:
```
// selected_status: string;
```

Comment (jmcmillenmusic) [reply apcd_cms/src/client/src/hooks/admin/index.ts]: Done and done!

File: apcd_cms/src/apps/utils/apcd_database.py [hunk 4]
```diff
@@ -152,6 +152,38 @@ def get_user_role(user):
...
+        users.user_email,
+        users.user_name,
+        submissions.payor_code
+        FROM submitter_users
+        JOIN users
+           ON submitter_users.user_id = users.user_id
+           AND submitter_users.user_number = users.user_number
+        JOIN submissions
+            ON submitter_users.submitter_id = submissions.submitter_id
+            ORDER BY users.user_id;
```
Comment (chandra-tacc) [inline apcd_cms/src/apps/utils/apcd_database.py:173]: submitter_users.user_id is in SELECT, so order by that one
```suggestion
            ORDER BY submitter_users.user_id;
```

**Review by chandra-tacc — APPROVED** [review]
LGTM. de-duping users and submitter users components can be done in another component.

**Pull request merged**